Auto generate E2E tests from Resnet50 and Resnext101

zhanglx13 opened the pull request:
As a first step to address [MLIR#619](https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/619), this PR does the following

- Use `-DROCK_E2E_TEST_ENABLED=0/1` to control whether the auto generated E2E tests are enabled. By "enabled" it means:
  - `ninja check-mlir-miopen` will generate E2E tests if they are enabled
  - The E2E tests is marked `unsupported` if they are disabled
- Create a new target `check-rock-e2e`, which generates and runs all E2E tests.
- Auto generate E2E tests using `generateE2ETest.py` and put them in `/build/mlir/test/e2e/`
- Move the test generation script `generateE2ETest.py` and config files `.toml` to `/mlir/test/e2e`, which defines a separate test suite named `ROCK E2E` (but it is still under the main test suite `MLIR MIOPEN`)

To check all tests as in CI, developer needs to set `-DROCK_E2E_TEST_ENABLED=1` before `ninja check-mlir-miopen` since `ROCK_E2E_TEST_ENABLED` is disabled by default.

To run the E2E tests only, developer can do `ninja check-rock-e2e`, which will regenerate the E2E tests if
- the local `lit.local.cfg` file is missing in `/build/mlir/test/e2e/` OR
- the config files `*.toml` files are changed

To force re-generate E2E tests, do `ninja re-gen-e2e-tests`.

To run all tests but the E2E tests, developer can only do `ninja check-mlir-miopen`. This is expected to accelerate the local development process.

If this framework works, we can move more E2E tests into the auto generation group.

zhanglx13 committed:
Auto generate E2E tests from Resnet50 and Resnext101

As a first step to address [MLIR#619](https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/619),
this PR does the following

- Use `-DROCK_E2E_TEST_ENABLED=0/1` to control whether the auto generated
E2E tests are enabled. By "enabled" it means:
  - `ninja check-mlir-miopen` will generate E2E tests if they are
enabled
  - The E2E tests is marked `unsupported` if they are disabled
- Create a new target `check-rock-e2e`, which generates and runs all E2E
tests.
- Auto generate E2E tests using `generateE2ETest.py` and put them in
`/build/mlir/test/mlir-miopen-driver/e2e/`
- Use a placeholder directory under `/mlir/test/mlir-miopen-driver/e2e`,
which defines a separate test suite named `ROCK E2E` (but it is still
under the main test suite `MLIR MIOPEN`)

To check all tests as in CI, developer can still do `ninja
check-mlir-miopen` without any changes since `ROCK_E2E_TEST_ENABLED` is
set to 1 by default.

To run the E2E tests only, developer can do `ninja check-rock-e2e`.

To run all tests but the E2E tests, developer needs to set
`ROCK_E2E_TEST_ENABLED` to 0 before running `ninja check-mlir-miopen`.
This is expected to accelerate the local development process.

If this framework works, we can move more E2E tests into the auto
generation group.

krzysz00 commented on mlir/test/CMakeLists.txt:
```diff
@@ -69,7 +69,14 @@ configure_lit_site_cfg(
+)
+
+configure_lit_site_cfg(
+  ${CMAKE_CURRENT_SOURCE_DIR}/mlir-miopen-driver/e2e/lit.site.cfg.py.in
```
Could we make the path just `e2e/`

krzysz00 commented on mlir/test/CMakeLists.txt:
```diff
@@ -117,17 +126,51 @@ if(MLIR_MIOPEN_DRIVER_ENABLED)
+
+# target to generate the E2E tests
+set(E2E_DIR ${CMAKE_CURRENT_BINARY_DIR}/mlir-miopen-driver/e2e)
+set(E2E_GEN_SCRIPT "${PROJECT_SOURCE_DIR}/mlir/utils/generateE2ETest.py")
```
We could perhaps move this out of `mlir/utils` and into `mlir/test/e2e`?

krzysz00 commented on mlir/test/CMakeLists.txt:
```diff
@@ -117,17 +126,51 @@ if(MLIR_MIOPEN_DRIVER_ENABLED)
+  COMMAND ${Python3_EXECUTABLE} ${E2E_GEN_SCRIPT} -d ${E2E_DIR} -i ${CONFIG_DIR}/Resnext101Config.toml
+)
+# Add a new target to run the E2E test only
+add_lit_testsuite(check-rock-e2e "Running rock e2e tests"
```
Why not move this into a `CMakeLists.txt` inside `e2e` and then have `if (ROCK_E2E_TESTS_ENABLED) add_subdirectory(e2e)`

krzysz00 commented on mlir/test/mlir-miopen-driver/e2e/lit.cfg.py:
```diff
@@ -0,0 +1,121 @@
+config.test_format = lit.formats.ShTest(not llvm_config.use_lit_shell)
+
+# suffixes: A list of file extensions to treat as test files.
+config.suffixes = ['.td', '.mlir', '.toy', '.ll', '.tc', '.py']
```
Given the restricted nature of the suite, maybe make this just `.mlir`?

krzysz00 commented on mlir/test/mlir-miopen-driver/e2e/lit.cfg.py:
```diff
@@ -0,0 +1,121 @@
+
+# test_source_root: The root path where tests are located.
+#config.test_source_root = os.path.dirname(__file__)
+config.test_source_root = os.path.join(config.mlir_obj_root, 'mlir/test/mlir-miopen-driver/e2e')
```
Same comment re why under mlir-miopen-driver?

krzysz00 commented on mlir/test/mlir-miopen-driver/e2e/lit.cfg.py:
```diff
@@ -0,0 +1,121 @@
+if config.target_triple:
+    config.available_features.add('default_triple')
+
+# Add the python path for both the source and binary tree.
```
I think we can drop this section

krzysz00 commented on mlir/test/mlir-miopen-driver/e2e/lit.cfg.py:
```diff
@@ -0,0 +1,121 @@
+config.environment['FILECHECK_OPTS'] = "-enable-var-scope --allow-unused-prefixes=false"
+
+
+# LLVM can be configured with an empty default triple
```
(and this section, he comments out of order)

krzysz00 commented on mlir/test/mlir-miopen-driver/e2e/lit.site.cfg.py.in:
```diff
@@ -0,0 +1,85 @@
+import sys
+import subprocess
+
+config.host_triple = "@LLVM_HOST_TRIPLE@"
```
I think we can drop a lot of these `config` keys

krzysz00 commented on mlir/test/mlir-miopen-driver/e2e/lit.site.cfg.py.in:
```diff
@@ -0,0 +1,85 @@
+# If rocm_agent_enumerator shows no viable GPUs, skip tests that need one,
+# because the default target will lead to compilation failures.
+config.no_AMD_GPU = False
+if config.rocm_path:
```
Note to future us: this is where the "only detect architecture once" goes

krzysz00 commented on mlir/utils/generateE2ETest.py:
```diff
@@ -117,6 +117,9 @@ def usage():
                     exclusions = generate_option_list(axis_prefixes, test, "exclude", "values")
                 for opt in combinations:
+                    # Only generate i8 data type for fwd convolutions
+                    if "i8" in opt[2] and ("conv2d_" in opt[0]):
```
Let's more precisely exclude `conv2d_bwd_weight` and `conv2d_bwd_data` or whatever the strings are

krzysz00 reviewed (commented):
I've got organizational nitpicks, but, them aside, this is a solid PR and a good step towards cleaning up our e2e test setup.

yiqian1 commented on mlir/test/mlir-miopen-driver/e2e/lit.cfg.py:
```diff
@@ -0,0 +1,121 @@
+    ], append_path=True)
+
+if not config.enable_rock_e2e_test or config.no_AMD_GPU:
+  config.unsupported = True
```
I don't think lit.cfg.py and lit.site.cfg.py.in are needed here. We can do this check in lit.local.cfg.

zhanglx13 replied:
Since the generated tests are in the build dir, do you mean also generate the `lit.local.cfg` file into the build dir with the tests?

yiqian1 replied:
I was thinking having these files in the parent folder and the local.cfg in e2e, but on the second thought, I agree with Krzysztof that we can just have e2e directly under build. There is no need to have a mlir-miopen-driver directory there.

yiqian1 commented on mlir/test/CMakeLists.txt:
```diff
@@ -117,17 +126,51 @@ if(MLIR_MIOPEN_DRIVER_ENABLED)
+  COMMENT "Generating rock E2E tests in ${E2E_DIR}"
+  COMMAND ${Python3_EXECUTABLE} ${E2E_GEN_SCRIPT} -d ${E2E_DIR} -i ${CONFIG_DIR}/Resnet50Config.toml
+  COMMAND ${Python3_EXECUTABLE} ${E2E_GEN_SCRIPT} -d ${E2E_DIR} -i ${CONFIG_DIR}/Resnext101Config.toml
+)
```
Can we make gen_e2e_tests depend on config files, i.e., re-generate only when the config files have been modified?

krzysz00 replied:
Yeah, good catch. And we probably want to stick the files in `mlir/test/e2e` too

yiqian1 replied:
Agree, moving the files to e2e makes more sense.

krzysz00 commented:
... Also, one big thing
Make `ROCK_E2E_TESTS` 0 by default
That way, CI doesn't run it by default
and patch nightly CI to set that variable

zhanglx13 committed:
Moved scripts and config to /mlir/test/e2e

zhanglx13 committed:
Only generate E2E tests when .toml files are changed

Also generate a local lit config to control whether the E2E tests are supported

zhanglx13 committed:
[jenkins] set ROCK_E2E_TEST_ENABLED in the same way as MIOPEN_DRIVER_E2E_TEST_ENABLED

krzysz00 commented on mlir/test/e2e/CMakeLists.txt:
```diff
@@ -0,0 +1,40 @@
+#   The local lit config file is missing OR
+#   Any of the *.toml files are changed.
+add_custom_command(OUTPUT ${LOCAL_CONFIG_FILE}
+  COMMAND echo "if not config.enable_rock_e2e_test or config.no_AMD_GPU:" > ${LOCAL_CONFIG_FILE}
```
I don't think this is it.
The LOCAL_CONFIG_FILE should be `COPY`ed in from the source directory.

But the other `DEPENDS` are reasonable

We might also have an `add_custom_command` that doesn't have an `OUTPUT`?

zhanglx13 replied:
We can keep the OUTPUT but use `copy` to generate the OUTPUT.

And when only the tests are missing (but `*.toml` and `lit.local.config` are not changed), they will not be re-generated. Do you want an option to force re-generate all tests or it is enough to only depend on the `*.toml` files and `lit.local.cfg`?

krzysz00 replied:
I figure that the tests should be generated if
1. They're not there (that is, the `lit.local.cfg` didn't get copied in) or
2. Someone's modified the config files

zhanglx13 replied:
what if only the tests are removed accidentally?

krzysz00 replied:
Good point. If that happens you either force the target to re-run or delete the `lit.local.cfg` in the output to force a regeneration.

krzysz00 replied:
Unless you don't like that

zhanglx13 replied:
I added a target `ninja re-gen-e2e-tests` to force generate the e2e tests.

krzysz00 reviewed (commented):
Minor comments, otherwise seems reasonable

zhanglx13 committed:
Copied lit.local.cfg from src and added a target to force generate e2e tests

krzysz00 commented on mlir/test/e2e/CMakeLists.txt:
```diff
@@ -0,0 +1,47 @@
+
+# Trigger the e2e tests generation unconditionally
+add_custom_target(re-gen-e2e-tests
+  DEPENDS __e2e_tests_regen
```
Do we know if that'll actually work? I don't think it'll regenerate if it thinks it doesn't need to. The better way for that would be `e2e_tests` with some sort of `--force`

zhanglx13 replied:
This works since its dependencies (`__e2e_tests_regen`) is always missing.

It seems that ninja does not provide a `--force`-like option to rebuild as mentioned [here](https://github.com/ninja-build/ninja/issues/317).
And we cannot associate the regeneration command with `add_custom_target` or `add_custom_command(TARGET)` since in this way the tests will always be regenerated.
So if you do not mind, I'll leave a TODO in the comment for future us to figure a better way to do this.

krzysz00 commented on mlir/test/e2e/CMakeLists.txt:
```diff
@@ -0,0 +1,47 @@
+#  1. conditionally if a target depends on ${LOCAL_CONFIG_FILE_BIN}.
+#  2. unconditionally if a target depends on __e2e_tests_regen since it is always missing.
+add_custom_command(OUTPUT ${LOCAL_CONFIG_FILE_BIN} __e2e_tests_regen
+  COMMAND cp ${LOCAL_CONFIG_FILE_SRC} ${LOCAL_CONFIG_FILE_BIN}
```
Probably fine for us, but I have some vague sense that CMake has its own `COPY` mechanism that's slightly more portable.

krzysz00 reviewed (approved):
Two nits, but approved

zhanglx13 committed:
Replaced cp with ${CMAKE_COMMAND} -E copy

krzysz00 commented:
No problem on leaving a TODO

Do you want to merge or should I?

zhanglx13 commented:
@krzysz00 I'll do it after I added the last changes and all CI tests pass.

zhanglx13 committed:
Grouped all configs in a list to make it easier to add more configs

krzysz00 reviewed (approved):
We might need to make sure MIOpen installs `tomli`, but on our end, this seems quite fine

krzysz00 commented:
(Unless I forgot what toml library we're using)

zhanglx13 commented:
It is `tomli` imported by `generateE2ETest.py`.

Why does MIOpen need this library? Will it call the test generation script?

krzysz00 commented:
... good point, they don't generate our tests, never mind all that
